motr/ut/idx_dix.c

```diff
@@ -782,15 +782,282 @@ struct m0_client* st_get_instance()
+	struct m0_reqh_service *duc_srv_svc;
+	struct m0_idx           duc_idx;
+	struct m0_container     duc_realm;
+	struct m0_fid           duc_ifid;
+};
+
+static struct dtm0_ut_ctx duc = {};
+
+static int duc_setup(void)
+{
+	const char              *cl_ep_addr   = "0@lo:12345:34:2";
+	struct m0_fid            cli_srv_fid  = M0_FID_INIT(0x7300000000000001,
+							    0x1a);
+	struct m0_fid            srv_dtm0_fid = M0_FID_INIT(0x7300000000000001,
+							    0x1c);
+	struct m0_reqh_service  *srv_srv = NULL;
+	struct m0_reqh          *srv_reqh;
+	int                      rc;
+	struct m0_container     *realm = &duc.duc_realm;
+
+	m0_fi_enable("m0_dtm0_in_ut", "ut");
+	rc = ut_suite_mt_idx_dix_init();
+	M0_UT_ASSERT(rc == 0);
+
+	m0_container_init(realm, NULL, &M0_UBER_REALM, ut_m0c);
+
+	if (ENABLE_DTM0) {
+		/* Connect to the server */
+		M0_UT_ASSERT(ut_m0c->m0c_dtms != NULL);
+		srv_reqh = &dix_ut_sctx.rsx_motr_ctx.cc_reqh_ctx.rc_reqh;
+		srv_srv = m0_reqh_service_lookup(srv_reqh, &srv_dtm0_fid);
+		rc = m0_dtm0_service_process_connect(srv_srv, &cli_srv_fid,
+						     cl_ep_addr, false);
+		M0_UT_ASSERT(rc == 0);
+	}
+
+	general_ifid_fill(&duc.duc_ifid, true);
+
+	/* Save the context */
+	duc.duc_cli_svc_fid = cli_srv_fid;
+	duc.duc_srv_svc_fid = srv_dtm0_fid;
+	duc.duc_srv_svc = srv_srv;
+
+	return 0;
+}
+
+static void idx_setup(void)
+{
+	struct m0_op            *op = NULL;
+	struct m0_idx           *idx = &duc.duc_idx;
+	struct m0_container     *realm = &duc.duc_realm;
+	int                      rc;
+	struct m0_fid           *ifid = &duc.duc_ifid;
+
+	m0_idx_init(idx, &realm->co_realm, (struct m0_uint128 *) ifid);
+
+	/* Create the index */
+	rc = m0_entity_create(NULL, &idx->in_entity, &op);
+	M0_UT_ASSERT(rc == 0);
+	m0_op_launch(&op, 1);
+	rc = m0_op_wait(op, M0_BITS(M0_OS_STABLE), WAIT_TIMEOUT);
+	M0_UT_ASSERT(rc == 0);
+	m0_op_fini(op);
+	m0_op_free(op);
+	op = NULL;
+}
+
+static void idx_teardown(void)
+{
+	struct m0_op            *op = NULL;
+	int                      rc;
+
+	/* Delete the index */
+	rc = m0_entity_delete(&duc.duc_idx.in_entity, &op);
+	M0_UT_ASSERT(rc == 0);
+	m0_op_launch(&op, 1);
+	rc = m0_op_wait(op, M0_BITS(M0_OS_STABLE), WAIT_TIMEOUT);
+	M0_UT_ASSERT(rc == 0);
+	m0_op_fini(op);
+	m0_op_free(op);
+	m0_idx_fini(&duc.duc_idx);
+	op = NULL;
+	M0_SET0(&duc.duc_idx);
+}
+
+
+static int duc_teardown(void)
+{
+	struct m0_reqh_service  *srv_srv = duc.duc_srv_svc;
+	struct m0_fid            cli_srv_fid  = duc.duc_cli_svc_fid;
+	int                      rc;
+
+	if (ENABLE_DTM0) {
+		/* Disconnect from the server */
+		rc = m0_dtm0_service_process_disconnect(srv_srv, &cli_srv_fid);
+		M0_UT_ASSERT(rc == 0);
+	}
+
+	rc = ut_suite_mt_idx_dix_fini();
+	m0_fi_disable("m0_dtm0_in_ut", "ut");
+	return rc;
+}
+
+/* Submits multiple M0 client (PUT|DEL) operations and then waits on EXECUTED,
+ * and then waits on STABLE.
+ */
+static void exec_then_stable(uint64_t nr, enum m0_idx_opcode opcode)
```

Comment:
exec_then_stable && exec_until_stable are copy/pasted. please reduce the amount of duplication.